### PR TITLE
Avoid null metric recorder remoteAddress parameters

### LINKED ISSUE #2748: java.lang.NullPointerException: Parameter specified as non-null is null: method a.b.c.ServerMetricsRecorder.recordDataReceived, parameter remoteAddress
### Summary
For URIs which URI parser is unable to parse, it throws exception which results in remote address being `null`. When `HttpMetricsRecorder` interface methods are invoked in Kotlin environment, it results in NPE.

### Code Flow
While resolving the path in `HttpServerOperations` object creation, if URI parser throws any exception due to parsing issue, remote address is coming as `null` due to `connectionInfo` being null.

## Expected Behavior
Metrics recorder should be invoked successfully. Interface methods params should be marked as nullable so that the implementation could provide specific handling. This would be specific to Kotlin as the arguments are inferred as non-nullable.

## Actual Behavior
Resulting in NPE as the interface methods params are non-nullable. 

## Steps to Reproduce

- Create an implementation of `ChannelMetricsRecorder`. 

- Use any URI which URI parser couldn't resolve. eg - `http://localhost:8080/s=set&_method=__construct&method=*&filter[]=system`

- When metrics recorder will invoke any of the interface methods requiring remote address, it will result in NPE in Kotlin.

## Possible Solution
Mark methods as nullable so that Kotlin can infer the parameters as nullable and implementation could handle those gracefully.

## Your Environment
Koltin Version: 1.6.10
reactor-netty-http: 1.0.18
* JVM version : JAVA 17
* OS and version : macOs Ventura 13.2

## Stack trace

**URI Parser**

`java.net.URISyntaxException: Illegal character in path at index 63: http://localhost:8080/s=set&_method=__construct&method=*&filter[]=system
	at java.net.URI$Parser.fail(URI.java:2974) ~[?:?]
	at java.net.URI$Parser.checkChars(URI.java:3145) ~[?:?]
	at java.net.URI$Parser.parseHierarchical(URI.java:3227) ~[?:?]
	at java.net.URI$Parser.parse(URI.java:3175) ~[?:?]
	at java.net.URI.<init>(URI.java:623) ~[?:?]
	at java.net.URI.create(URI.java:904) ~[?:?]
	at reactor.netty.http.HttpOperations.resolvePath(HttpOperations.java:389) ~[reactor-netty-http-1.0.18.jar!/:1.0.18]
	at reactor.netty.http.server.HttpServerOperations.<init>(HttpServerOperations.java:173) ~[reactor-netty-http-1.0.18.jar!/:1.0.18]
	at reactor.netty.http.server.HttpServerOperations.<init>(HttpServerOperations.java:148) ~[reactor-netty-http-1.0.18.jar!/:1.0.18]
	at reactor.netty.http.server.HttpTrafficHandler.channelRead(HttpTrafficHandler.java:213) ~[reactor-netty-http-1.0.18.jar!/:1.0.18]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) ~[netty-transport-4.1.76.Final.jar!/:4.1.76.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) ~[netty-transport-4.1.76.Final.jar!/:4.1.76.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) ~[netty-transport-4.1.76.Final.jar!/:4.1.76.Final]
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436) ~[netty-transport-4.1.76.Final.jar!/:4.1.76.Final]`

**Metrics recorder**

`io.netty.channel.AbstractChannelHandlerContext -- An exception 'java.lang.NullPointerException: Parameter specified as non-null is null: method com.apple.carrierservices.preauth.config.netty.HubbleHttpServerMetricsRecorder.incrementErrorsCount, parameter remoteAddress' [enable DEBUG level for full stacktrace] was thrown by a user handler's exceptionCaught() method while handling the following exception:
java.lang.NullPointerException: Parameter specified as non-null is null: method com.apple.carrierservices.preauth.config.netty.HubbleHttpServerMetricsRecorder.recordDataReceived, parameter remoteAddress
	at com.apple.carrierservices.preauth.config.netty.HubbleHttpServerMetricsRecorder.recordDataReceived(HubbleHttpServerMetricsRecorder.kt) ~[classes!/:?]
	at reactor.netty.http.server.AbstractHttpServerMetricsHandler.recordRead(AbstractHttpServerMetricsHandler.java:183) ~[reactor-netty-http-1.0.18.jar!/:1.0.18]
	at reactor.netty.http.server.AbstractHttpServerMetricsHandler.channelRead(AbstractHttpServerMetricsHandler.java:151) ~[reactor-netty-http-1.0.18.jar!/:1.0.18]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) ~[netty-transport-4.1.76.Final.jar!/:4.1.76.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) ~[netty-transport-4.1.76.Final.jar!/:4.1.76.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) ~[netty-transport-4.1.76.Final.jar!/:4.1.76.Final]
	at reactor.netty.http.server.HttpTrafficHandler.channelRead(HttpTrafficHandler.java:266) ~[reactor-netty-http-1.0.18.jar!/:1.0.18]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) ~[netty-transport-4.1.76.Final.jar!/:4.1.76.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) ~[netty-transport-4.1.76.Final.jar!/:4.1.76.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) ~[netty-transport-4.1.76.Final.jar!/:4.1.76.Final]
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436) ~[netty-transport-4.1.76.Final.jar!/:4.1.76.Final]`

### PR DESCRIPTION
**Pull request by pderop**
Motivation:

When a reactor-netty http server receives a request with an invalid request URI which can't be parsed, any registered metric recorders are invoked with some null `remoteAddress`parameters. Basically, when the URL can’t be parsed, a 400 bad request is sent using [HttpServerOperations.sendDecodingFailures](https://github.com/reactor/reactor-netty/blob/1.0.x/reactor-netty-http/src/main/java/reactor/netty/http/server/HttpServerOperations.java#L850)  which is using a [FailedHttpServerRequest](https://github.com/reactor/reactor-netty/blob/1.0.x/reactor-netty-http/src/main/java/reactor/netty/http/server/HttpServerOperations.java#L1023). And the constructor of FailedHttpServerRequest calls super(...) with null as the fifth parameter which is the connectionInfo.

That’s why [AbstractHttpServerMetricsHandler.recordWrite](https://github.com/reactor/reactor-netty/blob/1.0.x/reactor-netty-http/src/main/java/reactor/netty/http/server/AbstractHttpServerMetricsHandler.java#L254) passes a null remoteAddress parameter because ops.remoteAddress() returns null (connectionInfo is null).

Modification:

The ConnectionInfo object is now used when sending a decoding error. Now, if the ConnectionInfo could not be created in some rare conditions like when the Host or Forwarded/X-Forwarded-For headers are malformed, then we fallback on established connection info.

Fixes #2748

### CODE REVIEW

**Comment by pderop** [comment]
@violetagg , please wait a bit before checking, I may have to update the test before.
thanks.

**Comment by pderop** [comment]
it is ok.
I'm now waiting for the CI build results.

**Comment by pderop** [comment]
forced push after a rebase on top of 1.0.x branch, in order to pick up #2761

**Comment by pderop** [comment]
rebased on top of latest 1.0.x branch, in order to pick up #2787 and align to latest codebase version.

**Comment by pderop** [comment]
squashed all commits and rebased on top of latest 1.0.x branch.

File: reactor-netty-http/src/main/java/reactor/netty/http/server/HttpServerOperations.java [hunk 1]
```diff
@@ -826,6 +818,11 @@ static void sendDecodingFailures(
 							httpMessageLogFactory.warn(HttpMessageArgProviderFactory.create(msg)) : msg);
 		}
 
+		if (connectionInfo == null) {
```
Comment (violetagg) [inline reactor-netty-http/src/main/java/reactor/netty/http/server/HttpServerOperations.java:821]: `connectionInfo` is necessary only when we create `FailedHttpServerRequest`, isn't it?

Then at that point we have the original `request`

Why are checking for http/2?
`isHttp2 ? ctx.channel().parent() : ctx.channel()`

Comment (pderop) [reply reactor-netty-http/src/main/java/reactor/netty/http/server/HttpServerOperations.java:821]: > Why are we checking for http/2?
> isHttp2 ? ctx.channel().parent() : ctx.channel()

I just removed this useless check in commit https://github.com/reactor/reactor-netty/pull/2755/commits/e34cfc22d4b0e56b24ab05efb6c1eebdba13d4cd, sorry.

**Review by violetagg — APPROVED** [review]

**Comment by pderop** [comment]
@violetagg , thanks for the review !!

**Pull request merged**